codegen-llvm: never combine DSOLocal and DllImport

dvdhrm opened the pull request:
Prevent DllImport from being attached to DSOLocal definitions in the LLVM IR. The combination makes no sense, since definitions local to the compilation unit will never be imported from external objects.

Additionally, LLVM will refuse the IR if it encounters the combination (introduced in [1]):

```
  if (GV.hasDLLImportStorageClass())
    Assert(!GV.isDSOLocal(),
           "GlobalValue with DLLImport Storage is dso_local!", &GV);
```

Right now, codegen-llvm will only apply DllImport to constants and rely on call-stubs for functions. Hence, we simply extend the codegen of constants to skip DllImport for any local definitions.

This was discovered when switching the EFI targets to the static relocation model [2]. With this fixed, we can start another attempt at this.

[1] https://smlnj-gitlab.cs.uchicago.edu/manticore/llvm/commit/509132b368efed10bbdad825403f45e9cf1d6e38
[2] https://github.com/rust-lang/rust/issues/101656

dvdhrm committed:
codegen-llvm: never combine DSOLocal and DllImport

Prevent DllImport from being attached to DSOLocal definitions in the
LLVM IR. The combination makes no sense, since definitions local to the
compilation unit will never be imported from external objects.

Additionally, LLVM will refuse the IR if it encounters the
combination (introduced in [1]):

  if (GV.hasDLLImportStorageClass())
    Assert(!GV.isDSOLocal(),
           "GlobalValue with DLLImport Storage is dso_local!", &GV);

Right now, codegen-llvm will only apply DllImport to constants and rely
on call-stubs for functions. Hence, we simply extend the codegen of
constants to skip DllImport for any local definitions.

This was discovered when switching the EFI targets to the static
relocation model [2]. With this fixed, we can start another attempt at
this.

[1] https://smlnj-gitlab.cs.uchicago.edu/manticore/llvm/commit/509132b368efed10bbdad825403f45e9cf1d6e38
[2] https://github.com/rust-lang/rust/issues/101656

petrochenkov commented:
Could you add a `src/test/codegen` test case for this situation?
Static relocation mode can be enabled using `// compile-flags` even if it's not the default currently.
@rustbot author

petrochenkov commented:
>With this fixed, we can start another attempt at this.

Yeah, it would be nice to try reverting https://github.com/rust-lang/rust/pull/101413 when this is merged.

dvdhrm committed:
test/codegen: test inter-crate linkage with static relocation

Add a codegen-test that verifies inter-crate linkage with the static
relocation model. We expect all symbols that are part of a rust
compilation to end up in the same DSO, thus we expect `dso_local`
annotations.

dvdhrm commented:
@petrochenkov 
> Could you add a `src/test/codegen` test case for this situation?

I now added `static-relocation-mode.rs` to the codegen tests. For now, it *only* tests `dso_local` linkage in the static relocation model.

I would like to also add foreign linkage to this test, but currently *all* declarations are marked `dso_local` when the static relocation model is used. Is that intentional? That is, why can't I write a non-relocatable application which still uses dynamically linked dependencies?

dvdhrm commented:
@rustbot ready

petrochenkov commented:
@dvdhrm 
>I would like to also add foreign linkage to this test, but currently *all* declarations are marked `dso_local` when the static relocation model is used. Is that intentional? That is, why can't I write a non-relocatable application which still uses dynamically linked dependencies?

Apparently yes, and it's a relatively recent change (https://github.com/rust-lang/rust/pull/83592), but I didn't participate in that work and not sure whether the case you mention was considered.
cc @nagisa

petrochenkov commented on src/test/codegen/static-relocation-model.rs:
```diff
@@ -0,0 +1,25 @@
+    unsafe {
+        extern_decl::extern_fn() + extern_decl::extern_static
+    }
+}
```
I suspect that this test will fail on at least some targets on CI, and you'll have to use some per-target checking like in `src\test\codegen\abi-efiapi.rs`, or limit it to uefi.

petrochenkov commented:
@bors r+ rollup=iffy

bors commented:
:pushpin: Commit 46e0b025dc8a83d85fc2baef48079e9b7d5c45d2 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 46e0b025dc8a83d85fc2baef48079e9b7d5c45d2 -->
<!-- homu: {"type":"Approved","sha":"46e0b025dc8a83d85fc2baef48079e9b7d5c45d2","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Manishearth commented:
@bors r-

https://github.com/rust-lang/rust/pull/104791#issuecomment-1325796217

dvdhrm commented:
CI failed for MacOS, because there the `dso_local`-optimization is actually not as fine-grained, thus `dso_local` is not emitted for the test. I would expect other targets to have similar issues in the future, hence I decided to guard the test on `x86_64-pc-windows-msvc`. Sadly, I couldn't figure out how to run on windows *and* linux targets, but not anything else.

I accidentally also rebased on master, sorry for that! The only change from before is `# only-x86_64-pc-windows-msvc` in the test, and a rename from `static-relocation-model.rs` to `static-relocation-model-msvc.rs`.

@rustbot ready

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 7b9e1a95f0ca35bde8af529639dde501a0280425 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 7b9e1a95f0ca35bde8af529639dde501a0280425 -->
<!-- homu: {"type":"Approved","sha":"7b9e1a95f0ca35bde8af529639dde501a0280425","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

mati865 commented:
> CI failed for MacOS, because there the `dso_local`-optimization is actually not as fine-grained, thus `dso_local` is not emitted for the test. I would expect other targets to have similar issues in the future, hence I decided to guard the test on `x86_64-pc-windows-msvc`. Sadly, I couldn't figure out how to run on windows _and_ linux targets, but not anything else.
> 
> I accidentally also rebased on master, sorry for that! The only change from before is `# only-x86_64-pc-windows-msvc` in the test, and a rename from `static-relocation-model.rs` to `static-relocation-model-msvc.rs`.
> 
> @rustbot ready

How about `// ignore-macos` maybe combined with `// ignore-ios`?

bors commented:
:hourglass: Testing commit 7b9e1a95f0ca35bde8af529639dde501a0280425 with merge 8de4b138455add55bde6de5553a933a2ab79b71f...
<!-- homu: {"type":"BuildStarted","head_sha":"7b9e1a95f0ca35bde8af529639dde501a0280425","merge_sha":"8de4b138455add55bde6de5553a933a2ab79b71f"} -->

dvdhrm commented:
> How about `// ignore-macos` maybe combined with `// ignore-ios`?

Yeah, I thought about this as well, but as I mentioned I expect other targets to possibly have the same shortcuts in the `dso_local`-handling. Those might not be part of CI now, but might get upgraded to tier-2 at any point. I then decided to go with the `only-` marker because the way the `dso_local`-handling in the codegen-module (and llvm, fwiw) is written follows a similar logic. So new targets would get the same behavior as macos, not the same as windows or elf.

Additionally, I would like to test for `dllimport` in there as well, and that will definitely require a msvc-specific test. So maybe the better approach is to copy the test into `static-relocation-model-elf.rs` if we want to test ELF as well?

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3580285359/jobs/6024366681)
Approved by: petrochenkov
Pushing 8de4b138455add55bde6de5553a933a2ab79b71f to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3580285359/jobs/6024366681"},"merge_sha":"8de4b138455add55bde6de5553a933a2ab79b71f"} -->

rust-timer commented:
Finished benchmarking commit (8de4b138455add55bde6de5553a933a2ab79b71f): [comparison URL](https://perf.rust-lang.org/compare.html?start=d38a99078cd3c4dadac1ba8b729ea77e8d1d5a82&end=8de4b138455add55bde6de5553a933a2ab79b71f&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=d38a99078cd3c4dadac1ba8b729ea77e8d1d5a82&end=8de4b138455add55bde6de5553a933a2ab79b71f&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d38a99078cd3c4dadac1ba8b729ea77e8d1d5a82&end=8de4b138455add55bde6de5553a933a2ab79b71f&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.6% | [-2.1%, -1.1%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d38a99078cd3c4dadac1ba8b729ea77e8d1d5a82&end=8de4b138455add55bde6de5553a933a2ab79b71f&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

mati865 commented:
> Additionally, I would like to test for `dllimport` in there as well, and that will definitely require a msvc-specific test.

Seems more like Windows specific (so could be also tested on `windows-gnu`).

> So maybe the better approach is to copy the test into `static-relocation-model-elf.rs` if we want to test ELF as well?

If you think it would be beneficial...
